docs/dev-tools/debugging.md

```diff
@@ -0,0 +1,247 @@
+
+Outside of CiviCRM, the MySQL database software has features to enable the logging of all queries it receives.
+
+*TODO: how do we enable?*
+
+
+
+## Changing source code
+
+### In Smarty template files
+
+Add  `{debug}` to any part of the `.tpl` file and the Smarty Debug Window (described above) will display all variables in the same scope as the `{debug}` statement.
+
+### Printing PHP variables
+
+Show the contents of a variable:
+
+```php
+print_r($variable);
+```
+
+Show the contents of a variable, also with information regarding data types and lengths:
+
+```php
+var_dump($variable);
+```
+
+Another way to show the contents of a variable:
+
+```php
+CRM_Core_Error::debug($name, $variable = null, $log = true, $html= true);
+```
+
+Stop the script execution at that point.
+
+```php
+exit;
+```
+
+Print a backtrace:
+
+```php
+CRM_Core_Error::backtrace();
+```
+
+
+
+## Clearing the cache
+
+Clearing the cache is not a debugging technique, specifically. But sometimes it helps, and so is mentioned here for the sake of completeness.
+
+Using Drupal, you can clear all caches with the following `drush` command :
+
+-   `drush cc civicrm`
+-   `drush civicrm-cache-clear` *(older versions only)*
+
+Alternatively, you can call the following two methods:
+
+-   `CRM_Core_Config::clearDBCache();` clears the database cache
+-   `CRM_Core_Config::cleanup();` clears the file cache
+
+
+
+## Running a debugger program
+
+### What is a debugger?
+
+A debugger is a software program that watches your code while it
+executes and allows you to inspect, interrupt, and step through the
+code. That means you can stop the execution right before a critical
+juncture (for example, where something is crashing or producing bad
+data) and look at the values of all the variables and objects to make
+sure they are what you expect them to be. You can then step through the
+execution one line at a time and see exactly where and why things break
+down. It's no exaggeration to say that a debugger is a developer's best
+friend. It will save you countless hours of beating your head against
+your desk while you insert print statements everywhere to track down an
+elusive bug.
+
+Debugging in PHP is a bit tricky because your code is actually running
+inside the PHP interpreter, which is itself (usually) running inside a
+web server. This web server may or may not be on the same machine where
+you're writing your code. If you're running your CiviCRM development
+instance on a separate server, you need a debugger that can communicate
+with you over the network. Luckily such a clever creature already
+exists: XDebug.
+
+XDebug isn't the only PHP debugger, but it's the one we recommend for
+CiviCRM debugging.
+
+
+### Installing XDebug
+
+Debian / Ubuntu Linux
+
+```bash
+sudo apt-get install php5-xdebug
+```
+
+Red Hat / CentOS Linux
+
+```bash
+sudo yum install php-pecl* php-devel php-pear
+sudo pecl install Xdebug
+```
+
+Mac OS X
+
+```bash
+sudo port install php5-xdebug
```

Comment:
Thanks for pointing this out @totten. I added some of what your wrote in the commit below.